Add test case to test enum in has_many

jychen7 opened the pull request:
There is test in has_one to test enum, but there is no for has_many.

kamipo commented on activerecord/test/cases/associations/has_many_associations_test.rb:
```diff
@@ -377,6 +378,19 @@ def test_association_protect_foreign_key
+    author.my_books << book
+
+    assert_equal "reading", book.read_status
+    assert_not_equal 0, SpecialAuthor.joins(:my_books).where(books: { read_status: "reading" }).count
```
The reason that enum works is that `SpecialAuthor` inherites from `Author` and `has_many :books` is defined in the `Author`. This assertion should pass with defined by`class SpecialAuthor < ActiveRecord::Base`.

jychen7 replied:
I see, thx

I can make it RED now, using

```
  class SpecialAuthor < ActiveRecord::Base
    self.table_name = "authors"
    has_many :my_books, class_name: "SpecialBook", foreign_key: :author_id
  end

  class SpecialBook < ActiveRecord::Base
    self.table_name = "books"

    belongs_to :author
    enum read_status: { unread: 0, reading: 2, read: 3, forgotten: nil }
  end
```

still finding how to fix it

jychen7 replied:
https://github.com/rails/rails/blob/e2d8a2ccfd3b0d2030920346a2ff0cad93fcb188/activerecord/lib/active_record/relation/predicate_builder.rb#L69-L70

I think the problem is `associated_predicate_builder(books)` does not refer to `SpecialBook`, while `associated_predicate_builder(my_books)` does

jychen7 replied:
I can think of two way to fix it

1. keep using `SpecialAuthor.joins(:my_books).where(books: { read_status: "reading" })`
Expect: 
```
SELECT "authors".* FROM "authors" INNER JOIN "books" ON "books"."author_id" = "authors"."id" WHERE "books"."read_status" = 2
```
Actual: 
```
SELECT "authors".* FROM "authors" INNER JOIN "books" ON "books"."author_id" = "authors"."id" WHERE "books"."read_status" = 'reading'
```
Rails need to use table_name `books` go through all its associations to find out `my_books`, there maybe more associations using table `book`

2. add feature to support `SpecialAuthor.joins(:my_books).where(my_books: { read_status: "reading" })`
Expect: 
```
SELECT "authors".* FROM "authors" INNER JOIN "books" ON "books"."author_id" = "authors"."id" WHERE "books"."read_status" = 2
```
Actual: 
```
SELECT "authors".* FROM "authors" INNER JOIN "books" ON "books"."author_id" = "authors"."id" WHERE "my_books"."read_status" = 2
```
Rails need to use association `my_books` to find out table_name `books` and use it in sql

I would prefer way 2.
@kamipo what do you think?

kamipo replied:
way 1 is #30787.

way 2 had been applied but had already been reverted at 6a6c4c459106e2d9b76dd1233133a2bf30866ab0 to fix the regression #20308.

jychen7 replied:
@kamipo thanks for pointing that out.

for 6a6c4c4
`SpecialAuthor.joins(:books, :my_books)` will generate

```
SELECT "authors".* FROM "authors"
INNER JOIN "books" ON "books"."author_id" = "authors"."id"
INNER JOIN "books" "my_books_authors" ON "my_books_authors"."author_id" = "authors"."id"
```

and `where(books: { read_status: "reading" })` generates

```
WHERE "my_books"."read_status" = 2
```

is it possible to let `where` knows it should be `my_books_authors` from `joins`?

```
WHERE "my_books_authors"."read_status" = 2
```

kamipo replied:
It isn't possible at this point, since alias tracking for joins is lazy evaluated until `build_arel` is called.

jychen7 replied:
@kamipo thanks for the explanation. shall I update this PR to include new test cases for normal `has_many :books` only?

since there is no test to cover enum in `has_many`

jychen7 committed:
Add test case to test enum in has_many

There is test in has_one to test enum, but there is no for has_many.

[Rich Chen]
